### PR TITLE
[ML] throttle job audit msgs if delayed data occurs for consecutive buckets

### LINKED ISSUE #73679: [ML] Handle delayed data audit messages better
If a job is running a long time and continually gets delayed data, we write audit messages for each new bucket that gets delayed. 

For the annotations, we are able to update the previous annotation because we can keep track of the annotation ID there. But audit messages are traditionally append only, so we don't necessarily want to update the previous audit message. 

It may be good to introduce some sort of audit message "throttling" so that we don't write a new audit message if every new bucket has delayed data (maybe some multiple of bucket length with some min threshold)?

### PR DESCRIPTION
**Pull request by benwtrent**
In a consistently delayed environment, it is possible that many buckets
in a row have delayed data detected. This can flood the audit
message page, causing noise that detracts from other useful information

This commit reduces the number of audit messages written as the number
of consecutive buckets with delayed data increases.

closes https://github.com/elastic/elasticsearch/issues/73679

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedJob.java [hunk 1]
```diff
@@ -239,6 +253,16 @@ private void checkForMissingDataIfNecessary() {
         }
     }
 
+    boolean shouldWriteDelayedDataAudit() {
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedJob.java:256]:
```suggestion
    private boolean shouldWriteDelayedDataAudit() {
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedJob.java [hunk 2]
```diff
@@ -223,10 +224,23 @@ private void checkForMissingDataIfNecessary() {
                     && annotation.getEndTimestamp().equals(lastDataCheckAnnotationWithId.v2().getEndTimestamp())) {
                     return;
                 }
-
-                // Creating a warning in addition to updating/creating our annotation. This allows the issue to be plainly visible
-                // in the job list page.
-                auditor.warning(jobId, msg);
+                if (lastDataCheckAnnotationWithId != null) {
+                    long bucketAfterPreviousAnnotation = lastDataCheckAnnotationWithId.v2().getEndTimestamp().getTime()
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedJob.java:228]: Is it bucket start? bucket end? Could you rename this variable to convey its semantics better?

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedJob.java:228]: I agree this is a bit confusing. Is it safe to assume an annotation's end time is the start of the next bucket? If yes, could we compare the previous annotations end time with the current one's start time?

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedJob.java:228]: > with the current one's start time?

No, we cannot. In the case of consecutive buckets missing data, the start time stays at the first buckets start time.

What I could do is check the floor of `lastBucket` vs the endTime of `lastDataCheckAnnotationWithId`. That would indicate if this latest bucket overlaps with the endtime of the previous annotation.

**Review by przemekwitek — APPROVED** [review]
LGTM

**Pull request merged**